File: 112.path-sum.cpp
/*
 * @lc app=leetcode id=112 lang=cpp
 *
 * [112] Path Sum
 */
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    bool addNext(TreeNode* root, int sum, int target) {
        if(!root) return false;
        sum += root->val;
        if(sum == target && !root->left && !root->right) {
                return true;
        }
        return addNext(root->left, sum, target) || addNext(root->right, sum, target);

    }

    bool hasPathSum(TreeNode* root, int sum) {
        return addNext(root, 0, sum);
    }
};

File: 112.path-sum.py
#
# @lc app=leetcode id=112 lang=python3
#
# [112] Path Sum
#

# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def hasPathSum(self, root: TreeNode, sum: int) -> bool:
        if root == None:
            return False
        difference = sum - root.val
        if root.left == None and root.right == None:
            if difference == 0:
                return True
            else:
                return False
        
        return self.hasPathSum(root.left, difference) or self.hasPathSum(root.right, difference)

# @lc code=end

